File: AirRangingAPI/Domain/Models/Aircraft.cs
using System.ComponentModel.DataAnnotations;
using API.Domain.Models.Enums;

namespace API.Domain.Models
{
  public class Aircraft
  {
    // TODO: Validation
    [Key]
    public int Id { get; set; }
    
    [Required]
    [MaxLength(4)]
    public string IcaoId { get; set; }

    [Required]
    public string Manufacturer { get; set; }

    [Required]
    public string Model { get; set; }

    public string Variant { get; set; }
    public EAircraftType AircraftType { get; set; }
    public EEngineType EngineType { get; set; }
    public EWeightCategory WeightCategory { get; set; }
    public EIcaoWakeCategory IcaoWakeCategory { get; set; }
    public EFuelType FuelType { get; set; }

    public int MTOW { get; set; }

    public int CruiseSpeed { get; set; }

    [Required]
    public decimal FuelCapacity { get; set; }
    // TODO: Converted Fuel Weight <-> Volume

    [Required]
    public decimal MaxRange { get; set; }

    public int ServiceCeiling { get; set; }
  }
}

File: client/src/types/Aircraft/Aircraft.ts
import { PaginationInfo } from "../Pagination";
import { BaseModel, BaseModelSets, BaseModelSocials } from "../BaseModel";
import { UnionOmit } from "../UnionOmit";
import { EAircraftType, EEngineType, EFuelType, EIcaoWakeCategory, EWeightCategory } from "./AircraftEnums";

export interface Aircraft extends BaseModel {
  icaoId: string;
  manufacturer: string;
  model: string;
  variant?: string;
  registration?: string;
  aircraftType: EAircraftType;
  engineType: EEngineType;
  engineCount: number;
  weightCategory: EWeightCategory;
  icaoWakeCategory: EIcaoWakeCategory;
  fuelType: EFuelType;
  fuelCapacity: number;
  maxTakeoffWeight: number;
  minRunwayLength: number;
  cruiseSpeed: number;
  maxRange: number;
  serviceCeiling: number;
  enteredServiceAtYear: number;
}

export interface AircraftWithSocials extends Aircraft, BaseModelSocials{ }

export interface AircraftSelected extends AircraftWithSocials {
  loadedFuel: number;
  currentMaxRange: number;
}

export interface AircraftsSets {
  set: BaseModelSets;
}

export interface AircraftsFilterSearch extends AircraftsSets {
  set: BaseModelSets
  searchField: AircraftFieldsTypes;
  search: string;
}

export interface AdvancedFilters {
  advancedFilters: Array<AircraftsFilter>;
}

type AircraftsFilter = {
  field: AircraftFieldsTypes,
  value: any
};

export type AircraftWithoutIDs = UnionOmit<Aircraft, 'id' | "createdAtDate" | "modifiedAtDate">;

export type AircraftFieldsTypes = keyof AircraftWithSocials;

export type CloneAircraft = UnionOmit<AircraftWithSocials, "id" | "savesCount" | "authorUsername">;

export type AircraftSearchResult = {
  data: AircraftWithSocials[],
  pagination: PaginationInfo,
};
